Allow configuration of PlantUML includes (#207)
yuzutech/kroki: server/README.adoc

KROKI_NOMNOML_BIN_PATH:: Path to the `nomnoml` binary
KROKI_BLOCKDIAG_HOST:: The host of the blockdiag server
KROKI_BLOCKDIAG_PORT:: The port of the blockdiag server
KROKI_SAFE_MODE:: Either `SECURE` (default) or `UNSAFE`. Determines whether to sanitize requests for formats that allow including external content
KROKI_PLANTUML_INCLUDE_PATH:: The include path to set for PlantUML.

> KROKI_PLANTUML_INCLUDE_WHITELIST:: The name of a file that consists of a list or regexes for valid includes.

Reviewer: I guess that passing a list of regexes as environment variable is not practical but using a file means that you need to extend the docker image or bind mount a directory with this file right?
Is there an alternative solution?

Author: you could have one regex per env-var and chase up env vars with indexes, e.g. `KROKI_PLANTUML_INCLUDE_WHITELIST_0`, `KROKI_PLANTUML_INCLUDE_WHITELIST_1`, until you stop finding `KROKI_PLANTUML_INCLUDE_WHITELIST_n` for some integer n

KROKI_NOMNOML_BIN_PATH:: Path to the `nomnoml` binary
KROKI_BLOCKDIAG_HOST:: The host of the blockdiag server
KROKI_BLOCKDIAG_PORT:: The port of the blockdiag server
KROKI_SAFE_MODE:: One of:
+
--

> `SECURE`:: (default) Paranoid sanitization of requests before passing to image server.

Reviewer: I have used "diagram library" instead of "image server" in the documentation. I'm not a native speaker so maybe "image server" is more appropriated but in this case, we should replace every occurrences of "diagram library" otherwise it might be confusing.

Maybe we can stick with "diagram library" in this pull request and create another one to change the terminology if needed? What do you think?

Reviewer: So stick with "diagram library"?

Reviewer: ```suggestion
`SECURE`:: (default) Paranoid sanitization of requests before passing to the diagram library.
```

> `SAFE`:: Assume the image servers secure mode request sanitization is sufficient

Reviewer: ```suggestion
`SAFE`:: Assume the diagram libraries secure mode request sanitization is sufficient
```

> `UNSAFE`:: Run the image servers without request sanitization.

Reviewer: ```suggestion
`UNSAFE`:: Run the diagram libraries without request sanitization.
```

--
+
[NOTE]
--

> Some of the image servers allow referencing external entities by URL or accessing resources from the filesystem.

Reviewer: ```suggestion
Some of the diagram libraries allow referencing external entities by URL or accessing resources from the filesystem.
```

For example PlantUML allows the `!import` directive to pull fragments from the filesystem or a remote URL or the standard library.

It is the responsibility of the upstream codebases to ensure that they can be safely used without risk.

> Because Kroki does not perform code review of these servies, our default setting is to be paranoid and block imports unless known safe.

Reviewer: ```suggestion
Because Kroki does not perform code review of these services, our default setting is to be paranoid and block imports unless known safe.
```

> We encourage anyone running their own Kroki server to review the services security settings and select the security mode appropriate for their use case.

Reviewer: Not sure about this one:

```suggestion
We encourage anyone running their own Kroki server to review the services security settings and select the appropriated security mode for their use case.
```

Author: As a native English speaker, don’t make this change

Reviewer: Alright :baguette_bread: :fr:

--
KROKI_PLANTUML_INCLUDE_PATH:: The include path to set for PlantUML.

> KROKI_PLANTUML_INCLUDE_WHITELIST:: The name of a file that consists of a list or regexes for valid includes.

Reviewer: ```suggestion
KROKI_PLANTUML_INCLUDE_WHITELIST:: The name of a file that consists of a list of Java regular expressions for valid includes.
```

When one of the above library is requested, Kroki server will delegate the call.

==== PlantUML include files

> As PlantUML syntax includes pre-processor directives that have the potential to access external resources, serving live content from Kroki server is only recommended when using the dfeault settings of `KROKI_SAFE_MODE=SECURE` and `KROKI_PLANTUML_ALLOW_INCLUDE=false`.

Reviewer: ```suggestion
As PlantUML syntax includes pre-processor directives that have the potential to access external resources, serving live content from Kroki server is only recommended when using the default settings of `KROKI_SAFE_MODE=SECURE` and `KROKI_PLANTUML_ALLOW_INCLUDE=false`.
```

When one of the above library is requested, Kroki server will delegate the call.

==== PlantUML include files

As PlantUML syntax includes pre-processor directives that have the potential to access external resources, serving live content from Kroki server is only recommended when using the default settings of `KROKI_SAFE_MODE=SECURE` and `KROKI_PLANTUML_ALLOW_INCLUDE=false`.

If you are using Kroki server for offline image generation as part of your build tool chain, you may be able to change these defaults although it is not recommended.

The recommended pattern for handling includes is to either extend the docker image or bind mount a directory containing the files you wish to include and then set the `KROKI_PLANTUML_INCLUDE_PATH` environment variable.

For example:

[source,bash]
----
$ mkdir example-lib
$ printf "@startuml\nBob->Alice\n@enduml\n" > example-lib/bob.puml

> $ sudo docker run --publish=8000:8000 -v $(pwd)/example-lib:/example-lib -e KROKI_PLANTUML_INCLUDE_PATH=/example-lib yuzutech/kroki:latest

Author: ```suggestion
$ sudo docker run --publish=8000:8000 -v $(pwd)/example-lib:/example-lib -e KROKI_SAFE_MODE=SAFE -e KROKI_PLANTUML_INCLUDE_PATH=/example-lib yuzutech/kroki:latest
```
